KAFKA-13649: Implement early.start.listeners and fix StandardAuthorizer loading

cmccabe opened the pull request:
Since the StandardAuthorizer relies on the metadata log to store its ACLs, we need to be sure that
we have the latest metadata before allowing the authorizer to be used. However, if the authorizer
is not usable for controllers in the cluster, the latest metadata cannot be fetched, because
inter-node communication cannot occur. In the initial commit which introduced StandardAuthorizer,
we punted on the loading issue by allowing the authorizer to be used immediately. This commit fixes
that by implementing early.start.listeners as specified in KIP-801. This will allow in superusers
immediately, but throw the new AuthorizerNotReadyException if non-superusers try to use the
authorizer before StandardAuthorizer#completeInitialLoad is called.

For the broker, we call StandardAuthorizer#completeInitialLoad immediately after metadata catch-up
is complete, right before unfencing. For the controller, we call
StandardAuthorizer#completeInitialLoad when the node has caught up to the high water mark of the
cluster metadata partition.

This PR refactors the SocketServer so that it creates the configured acceptors and processors in
its constructor, rather than requiring a call to SocketServer#startup A new function,
SocketServer#enableRequestProcessing, then starts the threads and begins listening on the
configured ports. enableRequestProcessing uses an async model: we will start the acceptor and
processors associated with an endpoint as soon as that endpoint's authorizer future is completed.

Also fix a bug where the controller and listener were sharing an Authorizer when in co-located
mode, which was not intended.

cmccabe committed:
KAFKA-13649: Implement early.start.listeners and fix StandardAuthorizer loading [WIP]

Since the StandardAuthorizer relies on the metadata log to store its ACLs, we need to be sure that
we have the latest metadata before allowing the authorizer to be used. However, if the authorizer
is not usable for controllers in the cluster, the latest metadata cannot be fetched, because
inter-node communication cannot occur. In the initial commit which introduced StandardAuthorizer,
we punted on the loading issue by allowing the authorizer to be used immediately. This commit fixes
that by implementing early.start.listeners as specified in KIP-801. This will allow in superusers
immediately, but throw the new AuthorizerNotReadyException if non-superusers try to use the
authorizer before StandardAuthorizer#completeInitialLoad is called.

For the broker, we call StandardAuthorizer#completeInitialLoad immediately after metadata catch-up
is complete, right before unfencing. For the controller, we call
StandardAuthorizer#completeInitialLoad when the node has caught up to the high water mark of the
cluster metadata partition.

This PR refactors the SocketServer so that it creates the configured acceptors and processors in
its constructor, rather than requiring a call to SocketServer#startup A new function,
SocketServer#enableRequestProcessing, then starts the threads and begins listening on the
configured ports. enableRequestProcessing uses an async model: we will start the acceptor and
processors associated with an endpoint as soon as that endpoint's authorizer future is completed.

TODO: need to add an additional test case in QuorumControllerTest

hachikuji commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1864,6 +1780,18 @@ class ConnectionQuotas(config: KafkaConfig, time: Time, metrics: Metrics) extend
+    if (channel != null) {
+      debug(s"Closing connection from ${channel.socket.getRemoteSocketAddress}")
+      dec(listenerName, channel.socket.getInetAddress)
+      closeSocket(channel, this)
```
It's surprising to find this in `ListenerConnectionQuota`, also that we end up with a different logger. Is there any way we can pull it back to where it was? For example, maybe we could generalize `closeSocket`:

```scala
  def closeSocket(
    listenerName: ListenerName,
    channel: SocketChannel,
    connectionQuota: ConnectionQuotas,
    logging: Logging    
  )
```
Or maybe we can stick it into a trait so that we can get rid of the `logging` parameter.

cmccabe replied:
Well, the ConnectionQuotas class is tracking how many connections exist. It needs to be informed when a connection is closed. If you want, I can have the function take a Logging parameter so that we can log under `SocketServer` (or whatever) rather than `SocketServer.ConnectionQuotas`. But this is a DEBUG level message anyway... realistically, you don't ever see this in prod, so... not sure if it matters.

hachikuji replied:
Yes, I'm not disputing the need to register the close event with `ConnectionQuotas`. But I think it is a strange inversion of responsibility to have it close the socket as well. The `Processor` and `Acceptor` classes are the ones that own the socket lifecycle, while the `ConnectionQuotas` instance is just a parameter. The original implementation was more intuitive. 

It is admittedly a small thing, but I do think the logging issue matters. When we are debugging events in system test failures, for example, grepping the logger is a common strategy. It is surprising to need to include the connection quota object to see all of the close events.

cmccabe replied:
I guess to me, when you have something that is limiting the number of X objects, it makes sense for that thing to be logging about opening / closing the X objects as well. Then you can add helpful info about how many X objects there are, and so forth (in theory, I don't think we do that here).

So really this thing shouldn't be a "quotas" class but more like a pool or something. But we've changed enough in this PR, so let's not go there :)

Anyway, I will move it around, although we'll end up with some 4 or 5 argument utility functions.

cmccabe replied:
Update: I have changed it so that `closeChannel` takes a `Logging` argument (which will be either the Acceptor or Processor). This should lead to the log message about closure being listened under either the Acceptor or Processor, as it was before this PR.

hachikuji commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -1392,15 +1299,24 @@ private[kafka] class Processor(val id: Int,
+    if (shouldRun.getAndSet(false)) {
+      wakeup()
+      removeMetric("IdlePercent", Map("networkProcessor" -> id.toString))
+      metrics.removeMetric(expiredConnectionsKilledCountMetricName)
```
I know the code already does this, but metric removal seems like something to do after shutdown is complete. Maybe we can do it in `close`?

```scala
  def close(): Unit = {
    try {
      beginShutdown()
      thread.join()
    } finally {
      removeMetric("IdlePercent", Map("networkProcessor" -> id.toString))
      metrics.removeMetric(expiredConnectionsKilledCountMetricName)
    }
  }

hachikuji commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -681,24 +580,27 @@ private[kafka] abstract class Acceptor(val socketServer: SocketServer,
-      startProcessors(processors)
+  startFuture.thenRun(() => synchronized {
+    if (!shouldRun.get()) {
+      debug(s"Ignoring start future for ${endPoint.listenerName} since it has already been shut down.")
```
nit: "it" -> "the acceptor"?

hachikuji commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -104,184 +103,141 @@ class SocketServer(val config: KafkaConfig,
     }
-    info("Started socket server acceptors and processors")
+  })
+  newGauge(s"${ControlPlaneAcceptor.MetricPrefix}NetworkProcessorAvgIdlePercent", () => SocketServer.this.synchronized {
```
Do we register this metric even in kraft?

cmccabe replied:
Fair point. Let's skip these metrics when in kraft mode.

hachikuji commented on core/src/main/scala/kafka/network/SocketServer.scala:
```diff
@@ -104,184 +103,141 @@ class SocketServer(val config: KafkaConfig,
+  private var stopped = false
 
+  // Socket server metrics
+  newGauge(s"${DataPlaneAcceptor.MetricPrefix}NetworkProcessorAvgIdlePercent", () => SocketServer.this.synchronized {
```
The synchronization in these gauges seems less than ideal. Is it worth filing a jira to come up with a better approach?

cmccabe replied:
Yeah, it would be nice to avoid the synchronization. Filed KAFKA-13874 for this.

Dynamic listeners really make this kind of thing hard, even though very few people use them. :(

hachikuji commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -356,10 +359,13 @@ class BrokerServer(
 
-      /* Get the authorizer and initialize it if one is specified.*/
-      authorizer = config.authorizer
+      // Create and intiialize an authorizer if one is configured.
```
nit: typo "intiialize"

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -906,12 +906,35 @@ private void appendRaftEvent(String name, Runnable runnable) {
+        if (!needToCompleteAuthorizerLoad) return;
+        OptionalLong highWatermark = raftClient.highWatermark();
+        if (highWatermark.isPresent()) {
+            if (lastCommittedOffset + 1 >= highWatermark.getAsLong()) {
```
I guess the only issue with this is that the high watermark is a moving target. It probably works ok since writes to the metadata log should be infrequent. Not sure I have any better ideas. Maybe we could refresh the high watermark value only once every second or something like that.

cmccabe replied:
yeah, I think this is OK for now. if we have issues we can revisit...

cmccabe committed:
Address review comments, add test

cmccabe commented:
I made two changes just now:

1. Fixed it so TenantPartitionAssignor ignores fenced brokers, and added a unit test for this (as we discussed offline)

2. Removed ClusterMetadataTest#testNodeReplicaCountsForExistingTenantTopicWithInvalidAssignment and replaced it with  ClusterMetadataTest#testNodeReplicaCountsForUnregisteredBrokers. Basically, the assumptions of the old test were invalid (it is actually possible to have partitions placed on brokers that are no longer in the cluster).

Looking at the code, I see that it already took this into account when computing the nodeMetadata:

```
    String tenantPrefix = tenant + TenantContext.DELIMITER;
    for (Iterator<String> iterator = cluster.topicNames(); iterator.hasNext();) {
      String topic = iterator.next();
      boolean isTenantTopic = topic.startsWith(tenantPrefix);
      for (List<Integer> replicas : cluster.replicasForTopicName(topic)) {
        for (int i = 0; i < replicas.size(); i++) {
          Integer replica = replicas.get(i);
          NodeMetadata nodeMetadata = nodeMetadatas.computeIfAbsent(replica, node -> new NodeMetadata());
          nodeMetadata.incrementReplicas(i == 0, isTenantTopic);
        }
      }
    }
```

So the code was already handling this case smoothly, which is what we want. But the very artificial unit test (injecting a null into the replicas array, which wouldn't happen in a real Cluster object) was confusing the issue. Hopefully this clears up the issue.

hachikuji commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1486,13 +1486,16 @@ class KafkaConfigTest {
     assertEquals("3", originals.get(KafkaConfig.NodeIdProp))
   }
 
+  val kraftProps = new Properties()
```
nit: it would be a little nicer to make this a def. We don't need the properties after they are initialized.

hachikuji commented on metadata/src/main/java/org/apache/kafka/controller/QuorumController.java:
```diff
@@ -921,12 +925,35 @@ private void appendRaftEvent(String name, Runnable runnable) {
+        OptionalLong highWatermark = raftClient.highWatermark();
+        if (highWatermark.isPresent()) {
+            if (lastCommittedOffset + 1 >= highWatermark.getAsLong()) {
+                log.debug("maybeCompleteAuthorizerInitialLoad: completing authorizer " +
```
I think it would be reasonable to have an INFO message about this event. It only happens once and I suspect it will be useful to know when it completed.

hachikuji commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizer.java:
```diff
@@ -78,6 +77,24 @@ public AclMutator aclMutatorOrException() {
+        initialLoadFuture.complete(null);
+    }
+
+    // Visible for testing
```
nit: can probably make it default access?

cmccabe replied:
Unfortunately the junit integration test is in another package, so it must be public.

hachikuji commented on metadata/src/main/java/org/apache/kafka/metadata/authorizer/StandardAuthorizerData.java:
```diff
@@ -251,6 +273,9 @@ public AuthorizationResult authorize(
         if (superUsers.contains(principal.toString())) {
             rule = SuperUserRule.INSTANCE;
         } else {
+            if (!loadingComplete) {
```
nit: maybe pull this up one level?
```
} else if (!loadingComplete) {
  throw new AuthorizerNotReadyException();
} else {
...

hachikuji commented on core/src/test/scala/unit/kafka/server/KafkaConfigTest.scala:
```diff
@@ -1509,4 +1512,29 @@ class KafkaConfigTest {
     assertNotNull(config.getLong(KafkaConfig.SaslOAuthBearerJwksEndpointRetryBackoffMaxMsProp))
   }
+
+  @Test
```
May be useful testing that the default uses the controller listeners?

As a side comment, I wonder if the default of the controller listeners should only apply when the standard authorizer is in use? Or perhaps a warning of some kind if we have a non-standard authorizer and `early.start.listeners` is not set.

cmccabe replied:
> May be useful testing that the default uses the controller listeners?

Sure, I'll add a test.

> As a side comment, I wonder if the default of the controller listeners should only apply when the standard authorizer is in use? Or perhaps a warning of some kind if we have a non-standard authorizer and early.start.listeners is not set.

Hmm... it's not clear. If you're using the ZK-based authorizer, setting `early.start.listeners` neither helps nor hurts you.

In theory, anyone could write an `Authorizer` and implement this setting, so it's hard to warn about configurations that the hypothetical authorizer might or might not use. This is one of the usability taxes that all pluggable things impose on us...

hachikuji replied:
Hmm. I think the consequence is that the authorizer implementation has to be ready to receive authorization requests before the startup future has returned. At a minimum, I think we should update the documentation for `Authorizer.start` to mention the role of `early.start.listeners`.

cmccabe replied:
The authorizer does not have to be ready to receive early authorization requests unless it chooses to. For example, AclAuthorizer will return futures from `AclAuthorizer#start` that don't get completed early.

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -725,7 +727,11 @@ object KafkaConfig {
   val AuthorizerClassNameDoc = s"The fully qualified name of a class that implements <code>${classOf[Authorizer].getName}</code>" +
-  " interface, which is used by the broker for authorization."
+    " interface, which is used by the broker for authorization."
+  val EarlyStartListenersDoc = "A comma-separated list of listener names which should be started before any non-early start listeners. " +
```
nit: the first sentence seems circular. I think it would be helpful to mention the authorizer use case. Also, can we say something about the default?

cmccabe replied:
> nit: the first sentence is seems circular

Hmm, I don't know -- "started before any non-early-start listeners" does seem to convey some information. Circular would be more like "early start listeners are early start listeners," right? :) Maybe I'm misunderstanding your point.

> I think it would be helpful to mention the authorizer use case.

The second sentence says "This is useful in cases where the startup process requires some listeners to be open before other listeners can be brought up" which I thought was stating the use-case... do you have an idea for improving it?

> Also, can we say something about the default?

Good idea. I added a bit about the default (of controller listeners) at the end.

hachikuji replied:
The circularity was that we seemed to be defining the early start listeners in terms of the non-early start listeners. Here is my suggestion:

> A comma-separated list of listener names which should be started before the authorizer has finished initialization. This is useful when the authorizer is dependent on the cluster itself for bootstrapping, as is the case for the standard authorizer which stores ACLs in the metadata log. By default, all listeners included in `controller.listener.names` will also be early start listeners.

cmccabe replied:
OK. I took your text as the doc for early.start.listeners (with some minor changes)

hachikuji commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -1653,7 +1660,7 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
 
   /************* Authorizer Configuration ***********/
-  val authorizer: Option[Authorizer] = {
+  val createNewAuthorizer: Option[Authorizer] = {
```
This should be `def`. Otherwise, when running with `process.roles=broker,controller`, we will use the same `Authorizer` instance, which leads to duplicate errors.

cmccabe replied:
ugh. This explains some of the bugs we've seen, definitely. Yes, I'll fix that here...

cmccabe committed:
Address review comments

cmccabe committed:
Merge remote-tracking branch 'apache/trunk' into KAFKA-13649

cmccabe committed:
fix test

cmccabe committed:
update logging

cmccabe committed:
fix KafkaConfig and test etc

cmccabe committed:
Merge branch 'trunk' into KAFKA-13649

hachikuji reviewed (approved):
Thanks, LGTM.
